packages/gatsby-theme-docs/src/layouts/internals/sidebar.js

```diff
@@ -63,55 +67,166 @@ const LinkItem = styled.div`
+    }
+    // In case there was a scroll position saved in the location make sure that
+    // the scroll position is restored.
+    // We check for the active link to ensure that we scroll to the position only once.
+    else if (isActive && cachedScrollPosition >= 0) {
+      restoreScrollPosition();
+    }
+  }, [
+    linkRef,
+    cachedScrollPosition,
+    getChapterDOMElement,
+    restoreScrollPosition,
+  ]);
+
+  return (
+    <ClassNames>
+      {({ css: makeClassName }) => {
+        const linkClassName = makeClassName`
+          border-left: ${designSystem.dimensions.spacings.xs} solid ${designSystem.colors.light.surfaceSecondary1};
+          padding-left: calc(${designSystem.dimensions.spacings.m} - ${designSystem.dimensions.spacings.xs});
+          text-decoration: none;
+          color: ${designSystem.colors.light.textSecondary};
+          display: flex;
+          flex-direction: row;
+          align-items: flex-end;
+
+          :hover {
+            color: ${designSystem.colors.light.linkNavigation} !important;
+          }
+
+          > * + * {
+            margin: 0 0 0 ${designSystem.dimensions.spacings.s};
+          }
+        `;
+        const activeClassName = makeClassName`
+          border-left: ${designSystem.dimensions.spacings.xs} solid ${designSystem.colors.light.linkNavigation} !important;
+          color: ${designSystem.colors.light.linkNavigation} !important;
+        `;
+        return (
+          <Link
+            {...forwardProps}
+            innerRef={linkRef}
+            to={trimTrailingSlash(props.to)}
+            state={{
+              sidebarScrollPosition: nextScrollPosition,
+            }}
+            getProps={({ href }) => {
+              // Manually check that the link is the active one, even with trailing slashes.
+              // The gatsby link is by default configured to match the exact path, therefore we
+              // need to check this manually.
+              const linkPath = trimTrailingSlash(href);
+              if (linkPath === locationPath) {
+                return {
+                  className: [linkClassName, activeClassName].join(' '),
+                  'aria-current': 'page',
+                };
+              }
+              return { className: linkClassName };
+            }}
+          />
+        );
+      }}
+    </ClassNames>
+  );
+};
 SidebarLink.displayName = 'SidebarLink';
+SidebarLink.propTypes = {
+  to: PropTypes.string.isRequired,
+  nextScrollPosition: PropTypes.number.isRequired,
+  location: PropTypes.shape({
+    state: PropTypes.shape({
+      sidebarScrollPosition: PropTypes.number,
+    }),
+    pathname: PropTypes.string.isRequired,
+  }).isRequired,
+  getChapterDOMElement: PropTypes.func.isRequired,
+};
+
+const SidebarChapter = props => {
+  const elemId = `sidebar-chapter-${props.index}`;
+  const getChapterDOMElement = React.useCallback(
+    () => document.getElementById(elemId),
```

Comment:
FYI: previously we were using React `ref`, to access the DOM element. However, since we now use `useLayoutEffect`, the `ref.current` element is still not assigned, therefore we need to query the element on our own.
To make it easier, we define a unique ID and pass a callback function that returns the element for that ID.

Reply:
I have no emoptional feelings but am curious to understand what spoke against the approach I took using the DOM `Element.closest()` call to find the parent chapter?